File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/RegisterUser.java
package uk.co.victoriajanedavis.chatapp.domain.interactors;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.annotations.NonNull;
import uk.co.victoriajanedavis.chatapp.data.repositories.TokenRepository;
import uk.co.victoriajanedavis.chatapp.domain.entities.TokenEntity;
import uk.co.victoriajanedavis.chatapp.domain.interactors.ReactiveInteractor.SendInteractor;

public class RegisterUser implements SendInteractor<RegisterUser.RegisterParams, TokenEntity> {

    @NonNull
    private final TokenRepository repository;


    @Inject
    public RegisterUser(@NonNull final TokenRepository repository) {
        this.repository = repository;
    }

    @NonNull
    @Override
    public Single<TokenEntity> getSingle(@NonNull RegisterParams registerParams) {
        return repository.fetchTokenViaRegister(registerParams.username, registerParams.email,
                                                registerParams.password1, registerParams.password2)
                .andThen(repository.requestTokenSingle());
    }

    public static final class RegisterParams {
        @NonNull private final String username;
        @NonNull private final String email;
        @NonNull private final String password1;
        @NonNull private final String password2;

        public RegisterParams(String username, String email, String password1, String password2) {
            this.username = username;
            this.email = email;
            this.password1 = password1;
            this.password2 = password2;
        }
    }
}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/RegisterUser.kt
package uk.co.victoriajanedavis.chatapp.domain.interactors

import io.reactivex.Completable
import javax.inject.Inject

import uk.co.victoriajanedavis.chatapp.data.repositories.FirebaseTokenRepository
import uk.co.victoriajanedavis.chatapp.data.repositories.TokenRepository
import uk.co.victoriajanedavis.chatapp.domain.interactors.ReactiveInteractor.ActionInteractor

class RegisterUser @Inject constructor(
    private val backendTokenRepo: TokenRepository,
    private val firebaseTokenRepo: FirebaseTokenRepository
) : ActionInteractor<RegisterUser.RegisterParams> {

    // If POST of the Firebase registration token fails then whole registration process fails
    override fun getActionCompletable(params: RegisterParams): Completable {
        return backendTokenRepo.fetchTokenViaRegisterAndStore(
            params.username, params.email,
            params.password1, params.password2
        ).andThen(firebaseTokenRepo.postTokenToBackend()
            .onErrorResumeNext(::deleteTokenFromStorageThenError)
        )
    }

    private fun deleteTokenFromStorageThenError(e: Throwable): Completable {
        return backendTokenRepo.deleteTokenFromStorage()
            .andThen {
                Completable.error(Exception("Account created, but something went wrong logging in: ${e.message}"))
            }
    }

    class RegisterParams(
        val username: String,
        val email: String,
        val password1: String,
        val password2: String
    )
}
